static/js/ace-editor.ts

```diff
@@ -0,0 +1,337 @@
+      const whitespace = ace.require("ace/ext/whitespace");
+      whitespace.trimTrailingSpace(this._editor!.session, true);
+    } catch (e) {
+      console.error(e);
+    }
+  }
+
+  /**
+   * Resizes the editor after changing its size programatically
+   */
+  resize(): void {
+    this._editor?.resize()
+  }
+
+  /**
+   * Focuses the text area for the current editor
+   */
+  focus(): void { 
+    this._editor?.focus();
+  }
+
+  /**
+   * Clears the errors and annotations in the editor
+   */
+  clearErrors(): void {
+    // Not sure if we use annotations everywhere, but this was
+    // here already.
+    this._editor?.session.clearAnnotations();
+    this.markers?.clearErrors();
+  }
+
+  /**     
+   * Moves to the cursor to the end of the current file
+   */
+  moveCursorToEndOfFile(): void { 
+    this._editor?.navigateFileEnd();
+  }
+
+  /**
+   * Clears the selected text
+   */
+  clearSelection(): void {
+    this._editor?.clearSelection();
+  }
+
+  /**
+  * Removes all breakpoints on the rows.
+  **/
+  clearBreakpoints(): void { 
+    this._editor?.session.clearBreakpoints();
+  }
+
+  /**
+   * If this editor is used as a main editor, we set the options here
+   */
+  configureMainEditor(): void {
+    this._editor?.setShowPrintMargin(false);
+    this._editor?.renderer.setScrollMargin(0, 0, 0, 20)
+    this._editor?.addEventListener('change', () => {
+      theLocalSaveWarning.setProgramLength(this._editor!.getValue().split('\n').length);
+    });
+    // Set const value to determine the current page direction -> useful for ace editor settings
+    const dir = $("body").attr("dir");
+    if (dir === "rtl") {
+      this._editor?.setOptions({ rtl: true });
+    }
+
+    // If prompt is shown and user enters text in the editor, hide the prompt.
+    this._editor?.on('change', () => {
+      if (this.askPromptOpen) {
+        stopit();
+        this._editor?.focus(); // Make sure the editor has focus, so we can continue typing
+      }
+      if ($('#ask-modal').is(':visible')) $('#inline-modal').hide();
+      this.askPromptOpen = false;
+      $('#runit').css('background-color', '');
+      this.clearErrors();
+      //removing the debugging state when loading in the editor
+      stopDebug();
+    });
+
+    this._markers = new Markers(this._editor!);
+
+          // *** Debugger *** //
+      // TODO: FIX THIS
+      initializeDebugger({
+        editor: this._editor!,
+        markers: this.markers,
+        level: theLevel,
+        language: theLanguage,
+      });
+  }
+
+  set editor(editor: AceAjax.Editor) {
+    this._editor = editor;
+  }
+
+  /**
+ * The '@types/ace' package has the type of breakpoints incorrect
+ *
+ * It's actually a map of number-to-class. Class is usually 'ace_breakpoint'
+ * but can be something you pick yourself.
+ */
+  getBreakpoints(): Breakpoints {
+    return this._editor?.session.getBreakpoints() as unknown as Breakpoints;
+  }
```

Comment:
Same here, we probably need the functionality inside the class (because the `Breakpoints` class is also Ace specific)